### PR TITLE
Created model inspection notebook + fix code retrieval from samples

### PR DESCRIPTION
**Pull request by ian-coccimiglio**
This PR contains:
* [ ] a new test-case for the benchmark
  * [ ] I hereby confirm that NO LLM-based technology (such as github copilot) was used while writing this benchmark
* [ ] new dependencies in requirements.txt
  * [ ] The environment.yml file was updated using the command `conda env export > environment.yml`
* [ ] new generator-functions allowing to sample from other LLMs
* [ ] new samples (`sample_....jsonl` files)
* [ ] new benchmarking results (`..._results.jsonl` files)
* [x] documentation update
* [ ] bug fixes 

Related github issues (if relevant): #76 and #80 

Short description:
A notebook with 4 functions are provided to do post-hoc model inspection. They include functions for printing, filtering, assessing model performance, and translating the LLM completion strings into python functions in the local namespace (to allow local testing).

How do you think will this influence the benchmark results?
- None directly.

Why do you think it makes sense to merge this PR?
- These automated tools can be useful to verify the quality of the models and inspect their results.

### CODE REVIEW

File: demo/create_samples.ipynb
Comment (haesleinhuepf) [inline demo/create_samples.ipynb:114]: Might this modification break other LLM's benchmarks?

File: demo/retrieve_code_from_response.ipynb
Comment (haesleinhuepf) [inline demo/retrieve_code_from_response.ipynb:94]: I would be happy if there were no duplicate functions with almost the same name. That prevents confusion.

**Review by haesleinhuepf — CHANGES_REQUESTED** [review]
Hi @ian-coccimiglio ,

just a minor note: The recent changes have the potential to modiy the jsonl files. That's ok, if the PR title and description reflect that. Also sending two separate PRs for 1) the detection notebook, and 2) the modification of the code extraction would have been ok. Anyway, at this very moment I cannot merge the PR because if I executed the notebooks as they are now, they would not produce the content of the jsonl files. Let us know when this is ready for review / merge.

And thanks again for spending time on this!

Best,
Robert

**Comment by haesleinhuepf** [comment]
Hi @ian-coccimiglio ,

how are you doing? I would love to merge your code inspection notebook; but I would treat the updated code extraction stuff in a different branch and refine it a bit before merging. Would that be ok for you?

Thanks again!

Best,
Robert

**Comment by ian-coccimiglio** [comment]
Hey Robert, whatever works! Just got caught up in some ImageJ/Fiji work. If you're doing it today no worries, otherwise I'll clean up this PR tomorrow and make a separate one for the code extractions stuff.

**Comment by haesleinhuepf** [comment]
> I'll clean up this PR tomorrow and make a separate one for the code extractions stuff.

Awesome, thanks so much. And no hurry though. It's summer and weekend. Take your time ;-)

**Comment by ian-coccimiglio** [comment]
This PR is now just the model_inspection notebook. I don't /think/ it should modify the jsonl files, but let me know if I'm wrong about that.

File: demo/retrieve_code_from_response.ipynb
Comment (ian-coccimiglio) [reply demo/retrieve_code_from_response.ipynb]: Fixed in new PR, removing functions extract_python1, extract_python2, and consolidated to just "extract_python"

**Review by haesleinhuepf — APPROVED** [review]
Fantastic Ian, big thanks! That's very useful.

**Comment by haesleinhuepf** [comment]
one more question: @ian-coccimiglio If we put you in the author list, can you give me your full name and affiliation? Also do you have any conflicts of interest we need to report when submitting the paper?

**Comment by ian-coccimiglio** [comment]
> one more question: @ian-coccimiglio If we put you in the author list, can you give me your full name and affiliation? Also do you have any conflicts of interest we need to report when submitting the paper?

Hi Robert. Well, I'd hope to still add a few more things to the story! Nevertheless, if you end up adding me as an author, I wouldn't turn it down. Here's my name and affiliation.

Ian Francisco Coccimiglio
School of Biomedical Engineering, Faculty of Applied Science, University of British Columbia, Vancouver BC, Canada.

No conflicts of interest.

**Pull request merged**
